### ISSUE TITLE
Unusual characters in cmdline cause lines to disappear and corrupt the display

### ISSUE THREAD

**Issue by metayan** [nicolargo/glances#1692]
#### Description
When `cmdline` contains unusual characters, the `glances` display gets corrupted.
Complete lines go missing and some junk is shown on other lines.

For example when sorting on memory use, the processes in question are missing from the lower part of the display and only cause some junk to be displayed on other lines:
```
                                       0.0    0.3   4.13G 41.8M 70218 xxx            0:02 2    10 R /usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/Resources/Python.app/Contents/MacOS/Python /usr/local/bin/youtube-dl -v -w --write-description --write
                                       5.6    0.2   4.13G 40.0M 70267 xxx            0:09 2    10 R /usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/Resources/Python.app/Contents/MacOS/Python -m glances --fs-free-space --disable-check-update --percpu 

Accept-Ch                              0.0    0.2   4.24G 39.0M   488 xxx            0:08 4     0 R /System/Library/CoreServices/NotificationCenter.app/Contents/MacOS/NotificationCenter



Accept-Cha                             0.0    0.2   85.4G 37.6M  4043 xxx            0:04 5     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
                                       0.0    0.2   85.4G 37.6M   524 xxx            0:06 5     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent

Accept-Cha                             0.0    0.2   101G  36.8M   547 xxx            0:06 5     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
                                       0.0    0.2   101G  36.7M   615 xxx            0:07 5     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
                                       0.0    0.2   85.4G 36.7M   631 xxx            0:06 5     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
                                       0.0    0.2   85.4G 36.7M   598 xxx            0:06 5     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
                                       0.0    0.2   101G  36.7M   588 xxx            0:07 5     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
                                       0.0    0.2   101G  36.7M   603 xxx            0:06 5     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
```

When sorting on CPU use, causing the processes in question to be at the top of the process list, the left panel (network and disk info) is also corrupted.
```
                           0b     0b
en0                        0b     0b   CPU%   MEM%  VIRT  RES     PID USER          TIME+ THR  NI S Command
en1                        0b     0b   4.9    0.2   4.13G 40.1M 70267 xxx            0:12 2    10 R /usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/Resources/Python.app/Contents/MacOS/Python -m glances --fs-free-space --disable-check-update --percpu 
en2                        0b     0b   3.5    2.7   6.95G 438M    456 xxx           26:03 22    0 R /Applications/Safari.app/Contents/MacOS/Safari -psn_0_49164

Accept-Cha                 0b     0b   2.4    1.1   102G  175M  70170 xxx            0:10 11    0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
utun0                      0b     0b   1.8    12.6  6.37G 2.02G 14830 xxx         5h28:28 31    0 R /Applications/-edited
                                       1.7    0.7   86.0G 114M  14537 xxx           10:46 11    0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
DISK I/O                 R/s    W/s    1.4    0.6   102G  91.0M   609 xxx           14:34 8     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent


Accept-Cha               Free  Total   1.1    1.5   102G  250M  60338 xxx            0:32 13    0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
/ (disk1s1)             10.2G   447G   1.1    0.3   4.79G 47.8M   829 xxx           28:57 8     0 R /src/scm-edited




Accept-Cha                             0.7    1.3   85.9G 208M  51374 xxx            1:12 8     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
                                       0.6    0.1   4.22G 16.8M   448 xxx            2:20 44    0 R /usr/sbin/universalaccessd launchd -s
                                       0.5    0.2   4.25G 35.0M   463 xxx            4:12 4     0 R /System/Library/CoreServices/SystemUIServer.app/Contents/MacOS/SystemUIServer
                                       0.3    0.9   102G  141M  15849 xxx            2:47 16    0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
                                       0.3    0.6   5.52G 103M    525 xxx            4:52 14    0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.Networking.xpc/Contents/MacOS/com.apple.WebKit.Networking
                                       0.2    0.5   85.8G 87.0M   611 xxx            2:02 9     0 R /System/Library/StagedFrameworks/Safari/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
                                       0.2    0.1   4.18G 23.8M   494 xxx            1:24 3     0 R /usr/libexec/nsurlstoraged
```

Some of the culprit processes as shown by `ps auxSww` (with links slightly shortened):
```
xxx              69959   3.1  0.2  4430132  36516 s016  SN    8:14AM   0:00.39 ffmpeg -y -loglevel verbose -headers User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.96 Safari/537.36^M\012Accept-Charset: ISO-8859-1,utf-8;q=0.7,*;q=0.7^M\012Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8^M\012Accept-Encoding: gzip, deflate^M\012Accept-Language: en-us,en;q=0.5^M\012 -i https://ada.net/exp=15~acl=%2Ff955%2F%2A~hmac=338f62c1/playlist.m3u8 -c copy -f mp4 file:master-master.faudio-high-audio.mp4.part
xxx              69965   2.5  0.2  4421876  35404 s016  SN    8:14AM   0:00.25 ffmpeg -y -loglevel verbose -headers User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3714.0 Safari/537.36^M\012Accept-Charset: ISO-8859-1,utf-8;q=0.7,*;q=0.7^M\012Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8^M\012Accept-Encoding: gzip, deflate^M\012Accept-Language: en-us,en;q=0.5^M\012 -i https://ada.net/exp=15~acl=%2F5d07%2F%2A~hmac=7d963177/playlist.m3u8 -c copy -f mp4 file:master-master.faudio-high-audio.mp4.part
xxx              69899   2.3  0.2  4440168  36888 s016  SN    8:13AM   0:00.70 ffmpeg -y -loglevel verbose -headers User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3632.5 Safari/537.36^M\012Accept-Charset: ISO-8859-1,utf-8;q=0.7,*;q=0.7^M\012Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8^M\012Accept-Encoding: gzip, deflate^M\012Accept-Language: en-us,en;q=0.5^M\012 -i https://ada.net/exp=15~acl=%2Fcf6758%2F%2A~hmac=fd7b0f3f/playlist.m3u8 -c copy -f mp4 file:master-master.f746.mp4.part

```

#### Versions
Glances v3.1.5_DEVELOP with PsUtil v5.7.0
(actually v3.1.4-19-gb0dc3ff5)
macOS 10.13.6 (17G12034)

#### Logs
Started and stopped `glances` several times to try to figure out what was causing the display corruption. A typical run looks like this:
```
2020-07-13 08:15:35,679 -- INFO -- Start Glances 3.1.5_DEVELOP
2020-07-13 08:15:35,679 -- INFO -- CPython 3.7.7 and psutil 5.7.0 detected
2020-07-13 08:15:35,684 -- INFO -- Search glances.conf file in /Users/xxx/Library/Application Support/glances/glances.conf
2020-07-13 08:15:35,684 -- INFO -- Search glances.conf file in /usr/local/etc/glances/glances.conf
2020-07-13 08:15:35,687 -- INFO -- Read configuration file '/usr/local/etc/glances/glances.conf'
2020-07-13 08:15:35,689 -- WARNING -- PsUtil can not grab processes io_counters ('Process' object has no attribute 'io_counters')
2020-07-13 08:15:35,698 -- INFO -- Start GlancesStandalone mode
2020-07-13 08:15:35,789 -- WARNING -- Missing Python Lib (No module named 'netifaces'), IP plugin is disabled
2020-07-13 08:15:35,798 -- WARNING -- Error loading Docker Python Lib. Docker plugin is disabled (No module named 'docker')
2020-07-13 08:15:35,799 -- ERROR -- docker plugin - Can not connect to Docker (name 'docker' is not defined)
2020-07-13 08:15:35,800 -- WARNING -- Missing Python Lib (No module named 'pySMART'), HDD Smart plugin is disabled
2020-07-13 08:15:35,803 -- WARNING -- Missing Python Lib (No module named 'pymdstat'), Raid plugin is disabled
2020-07-13 08:15:35,810 -- WARNING -- Sparklines module not found (No module named 'sparklines')
2020-07-13 08:15:35,812 -- WARNING -- Missing Python Lib (No module named 'cpuinfo'), Quicklook plugin will not display CPU info
2020-07-13 08:15:35,813 -- WARNING -- Missing Python Lib (No module named 'py3nvml'), Nvidia GPU plugin is disabled
2020-07-13 08:15:35,818 -- WARNING -- Missing Python Lib (No module named 'wifi'), Wifi plugin is disabled
2020-07-13 08:15:35,818 -- WARNING -- Wifi lib is not compliant with Python 3, Wifi plugin is disabled
2020-07-13 08:15:55,059 -- INFO -- Stop Glances (keypressed: 113)
```

**Comment by nicolargo** [comment]
Hi @metayan 

Thanks for the bug report.

Patch is available in the develop branch (commit 9f8a8120).

Can you test it ?

**Comment by metayan** [comment]
Same behavior as before.

Tested on Ubuntu 18.04 as well, and there only the first line of the multiline command is shown in `glances` (line breaks added here):
```
0.0    1.2   518M  48.0M 22969 xxx            0:06 1     0 S    0 0    
ffmpeg -y -headers User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36
 (KHTML, like Gecko) Chrome/72.0.3609.2 Safari/537.36
```
This works even with the version before commit https://github.com/nicolargo/glances/commit/9f8a812081dc30ed17b387f9b47740cfd4272e99.


Did some more testing, and **if `cmdline` contains several lines, then the lines contain `\n` on Linux, but `\r\n` on macOS**, but this is probably determined by the calling program, in this case `youtube-dl`.

#### To see what's happening I added this (only for debugging):
```diff
diff --git a/glances/plugins/glances_processlist.py b/glances/plugins/glances_processlist.py
index 47029f40..ff061766 100644
--- a/glances/plugins/glances_processlist.py
+++ b/glances/plugins/glances_processlist.py
@@ -321,6 +321,8 @@ class Plugin(GlancesPlugin):
         else:
             cmdline = '?'
         try:
+            msg = self.layout_stat['command'].format(cmdline) + os.sep
+            ret.append(self.curse_add_line(msg, splittable=True))
             if cmdline:
                 path, cmd, arguments = split_cmdline(cmdline)
                 if os.path.isdir(path) and not args.process_short_name:
```

#### and it revealed (just before `Accept-Charset`):
macOS:
```
['ffmpeg', '-y', '-headers', 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36
 (KHTML, like Gecko) Chrome/74.0.3729.12 Safari/537.36\r\nAccept-Charset...
```

Linux:
```
['ffmpeg', '-y', '-headers', 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36
 (KHTML, like Gecko) Chrome/72.0.3602.0 Safari/537.36\nAccept-Charset...
```

Probably the most general solution is to replace all control characters with something like `' '`, `'_'` or `'?'` before sending them to the screen.
